### PR TITLE
KAFKA-6687: rewrite topology to allow reading the same topic multiple times in the DSL

### PR DESCRIPTION
**Pull request by ableegoldman**
Needed to fix this on the side in order to more easily set up some experiments, so here's the PR.

Allows a user to create multiple KStreams ~and/or KTables~ from the same topic, collection of topics, or pattern. At the moment this isn't possible since we can only consume from a topic once, and each source topic maps to a single source node in the topology. The "fix" is just to rewrite the logical plan and merge any duplicate source nodes into a single node before it gets compiled into the physical topology.

The one exception is when the stream/table are subscribed to an overlapping-but-unequal collection of topics, which I left as future work (with a TODO in the comments describing a possible solution). If the offset reset policy doesn't match we just throw a TopologyException.

edit: tables are much more complicated so I opted to restrict things to just multiple KStreams for now on and consider allowing multiple KTables (or KStream+KTable) as followup work

### CODE REVIEW

**Comment by ableegoldman** [comment]
Call for review @mjsax @cadonna @lct45

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:424]: Saw this and at first I thought it was broken because it only considers pattern-subscribed topics that happened to explicitly configure an offset reset policy. Unless I'm missing something here, that makes no sense and we should consider _all_  source patterns and whether they overlap.
But then I started thinking, why does it matter if they overlap? Just because one pattern is a substring of another does not mean that they'll match the same topics. So I think that we should actually just remove this restriction altogether. Am I missing anything here?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: Could you please add a try-catch clause to better document the test?
For example:
```suggestion
    public void shouldAllowReadingFromSameTopic() {
        builder.stream("topic");
        builder.stream("topic");
        
        try {
            builder.build();
        } catch (final TopologyException topologyException) {
            fail("TopologyException not expected");
        }
    }
```
This applies also to the other tests.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:331]: We could avoid the `instanceof` and the casting if we introduce a `RootGraphNode` with a method `sourceNodes()`. Since a root can only have source nodes and state stores as children, we could make the topology code in general a bit more type safe. As far as I can see that would need some additional changes outside the scope of this PR. So, feel free to not consider this comment for this PR and we can do another PR for that.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:327]: Just to be clear. This improves the situation but it is not a complete solution, right? Assume we have a topic `topicA`. Patterns `topic*` and `topi*` both  match `topicA` but they are different when compared with this comparator. In that case a `TopologyException` would be thrown in the `InternalTopologyBuilder`, right?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java]: Do we really need this comment and the comment on line 81. We get the same information when we navigate to the call and to the implementation of the methods with the difference that comments can start to be outdated without us noticing it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:424]: I agree on the first part.
Regarding the second part, I had similiar thoughts when I wrote my comment in `mergeDuplicateSourceNodes()`.
But I might also be missing something here.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: Please use a collection with at least two topics to test the loop over the collections.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java]: What should happen if this reset policy is `null` and the other is not `null`? I guess we should also throw in that case, don't we?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:956]: What should happen in this case? See also my comment in `merge()`.
```
    @Test
    public void shouldThrowWhenSubscribedToATopicWithSetAndUnsetResetPolicies() {
        builder.stream("topic");
        builder.stream("topic", Consumed.with(AutoOffsetReset.EARLIEST));
        assertThrows(TopologyException.class, builder::build);
    }
```

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:971]: Could you also add a test with two patterns with the same string but one with a set reset policy and one with unset reset policy like you did for the non-pattern case. Just to make it clear it should also throw in that case.

**Review by cadonna — COMMENTED** [review]
Thank you for the PR, @ableegoldman!

According to the ticket similar work needs to be done for `table()` and `globalTable()`. What do you think of adding subtasks to the ticket to track what has already been done and what not?

Here my feedback.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java]: I'll remove it

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java]: Ack, good catch

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:956]: Yep that was just an oversight in the condition in `merge()`. I fixed that and added another unit test for the case

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:331]: Yeah I think that's a fair point but I would prefer to keep the scope of this PR as small as possible for now. Maybe @lct45 could pick this up on the side once this is merged?

**Comment by ableegoldman** [comment]
> According to the ticket similar work needs to be done for table() and globalTable(). What do you think of adding subtasks to the ticket to track what has already been done and what not?

Thanks for the review @cadonna . This fix actually does work for `table()` as well as `stream()`, I've updated the title and added unit tests to reflect this. As for the global table, there is already a separate ticket for this which should be linked to from KAFKA-6687.

I do think there's some possible followup work to further improve the situation (see my comment above) but I would say that's it's different enough to merit creating separate followup tickets rather than subtasks of this one. Lmk what you think

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:327]: Yes to all of that: this PR improves some situations, but not all. Specifically you would still get a TopologyException if  (a) subscribing to overlapping but not equal collection of topics, (b) subscribing to a topic and to a pattern that matches said topic, and (c) subscribing to two (or more) patterns that match the same topic(s).

Case (c) is what you described, I just wanted to list them all here for completion. Here's my take on what we can/should reasonably try to tackle:

(a) this case is easily detected, easily worked around, and easy for us to fix. It results in a "compile time" exception (meaning when the topology is compiled, not the program) which users can quickly detect and work around if need be by rewriting the topology themselves. Fix is relatively straightforward but very low priority, so I plan to just file a followup ticket for this for now
(b) is easily detected (you get a compile time exception) and possible to work around, but difficult to solve. I think in all cases a user could find a way around this issue by some combination of topology rewriting and Pattern manipulation or topic renaming, depending on what exactly they're trying to achieve. Of course there's no way for us to detect what an arbitrary user is trying to do in this case, so I don't see any path forwarding to making this case possible. No plans to file a followup ticket
(c) is difficult to detect, might be possible to work around, and probably very complicated to actually fix. Unfortunately, in this case you only get a run-time exception, since there's no way of knowing which topics will or will not be created ahead of time. And I'm thinking that determining whether two regexes will both match any possible string may be unsolvable...so, no followup ticket planned for this. 
WDYT?

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:327]: Thank you for the list of issues. I agree in all points.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java]: I would also remove this comment.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: Could you please extract this part in its own method since we use it also in a couple of other tests?

**Comment by cadonna** [comment]
> I do think there's some possible followup work to further improve the situation (see my comment above) but I would say that's it's different enough to merit creating separate followup tickets rather than subtasks of this one. Lmk what you think

Fair enough. Let's do it as you say.

**Review by cadonna — APPROVED** [review]
LGTM!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (lct45) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:331]: SGTM

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (lct45) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:424]: I might _also_ be missing something, but what's the scenario where one pattern is a substring of another and they _dont_ match the same topics? If you take Bruno's example from earlier of `topic*` and `topi*`, `topi*` would be considered a substring of `topic*` and they would both match `topic A`, right? I guess the other scenario is if we have a topic `topia A`, that would match `topi*` and not `topic*`. So I guess it seems like it isn't always true that they'll overlap, but we would want to check if they do, right?

**Review by lct45 — APPROVED** [review]
LGTM pending Bruno's final comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:424]: Pattern `topic*` is contained in pattern `topic*A`. However, `topic*A` matches only a subset of `topic*`. So, they do not match exactly the same topics. But matching exactly the same topics is a pre-requisite for merging the source nodes.

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:424]: I think in this case we were matching whether the pattern's string was a literal substring of another pattern's string, not whether the regexes themselves are substrings. So `topi*` would not be a substring of `topic*` because `topi*` is not contained literally within the string `topic*`. It's not doing a smart regex-matching, just a dumb  literal string comparison

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:349]: Filed https://issues.apache.org/jira/browse/KAFKA-10721

**Comment by ableegoldman** [comment]
Two unrelated flaky test failures:
`EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[true]`
`DynamicConnectionQuotaTest.testDynamicListenerConnectionCreationRateQuota`

**Comment by ableegoldman** [comment]
Merged to trunk

**Pull request merged**
